CorsixTH/Lua/entities/machine.lua

```diff
@@ -93,91 +96,130 @@ local function setSmoke(self, isSmoking)
-        num_extinguishers = num_extinguishers + 1
-      end
-    end
-    if num_extinguishers == 0 or threshold < -3 then
-      -- If no extinguisher in room, or machine used 5 times over its strength always explode
-      explode = true
-    else
-      -- Explosion chance increases 20% with every use over strength, and reduced by 5% for every additional extinguisher (up to 3 extra) in the room bar the first one
-      explosion_chance = (2 / self.strength) + (threshold * -0.2) - (num_extinguishers * 0.05) + 0.05
-      -- Cap it until guaranteed explosion
-      explosion_chance = explosion_chance > 0.95 and 0.95 or explosion_chance
-      explosion_chance = explosion_chance < 0.05 and 0.05 or explosion_chance
-      explode = math.random() < explosion_chance
-    end
-  end
-  -- Room failed to be saved, or no extinguishers were present
-  if explode then
-    -- Clean up any task of handyman coming to repair the machine
-    self.hospital:removeHandymanTask(taskIndex, "repairing")
-    -- Blow up the room
-    room:crashRoom()
-    self:setCrashedAnimation()
-    -- No special cursor required when hovering over the crashed room
-    self.hover_cursor = nil
-    -- Clear dynamic info (tracks machine usage which is no longer required)
-    self:clearDynamicInfo()
-    -- Prevent the machine from smoking, it's now just a pile of rubble
-    setSmoke(self, false)
-    -- If we have the window for this machine open, close it
-    local window = self.world.ui:getWindow(UIMachine)
-    if window and window.machine == self then
-      window:close()
-    end
-    -- Clear the icon showing a handyman is coming to repair the machine
-    self:setRepairing(nil)
+  local remaining_use_count = self:getRemainingUses()
+  local repair_task_index = self.hospital:getIndexOfTask(self.tile_x, self.tile_y, "repairing")
+  local must_explode = self:calculateIsMachineMustExplode(room, remaining_use_count)
+  if must_explode then
+    -- Room failed to be saved, it must be explode
+    self:explodeMachine(room)
     return true
-  -- Else if urgent repair needed or room didn't explode
-  elseif threshold < 4 then
+  else
+    -- Else if urgent repair needed or room didn't explode
+    self:checkIsNeedToCallHandymanForRepair(room, remaining_use_count, repair_task_index)
+    -- Update whether smoke gets displayed for this machine (and if so, how much)
+    self:calculateSmoke(room)
+  end
+end
+
+--! Call on machine used. After machine used increment use values accordingly.
+function Machine:incrementUsedCount()
+  self.times_used = self.times_used + 1
+  self.total_usage = self.total_usage + 1
+end
+
+--! Call on machine use. Handles crashing the machine & queueing repairs
+--!param room (object) machine room
+--!param remaining_use_count (int) machine remaining use count
+--!param repair_task_index (int) task id of repairing this machine handyman job
+function Machine:checkIsNeedToCallHandymanForRepair(room, remaining_use_count, repair_task_index)
```

Comment:
I don't see any machine crashing handling here, as claimed by line 138.
It just handles handyman tasks for repair.

The function name claims to "check". To me that means it returns a check result and doesn't actually do anything. This function not only checks, it also actively handles calling a handyman. The verb "check" is incomplete thus.

I find the `remaining_use_count` parameter a bit weird. This is the `Machine` class, why doesn't the machine already know this?

Also not sure about the repair_task_index. Can't you find that out? The machine knows its hospital as far as I know. It also knows its position, I hope.